File: 16. 3Sum Closest.py
class Solution(object):
    def threeSumClosest(self, nums, target):
        """
        :type nums: List[int]
        :type target: int
        :rtype: int
        """
        res = target
        min_diff = 1<<29
        i, ln = 0, len(nums)
        nums.sort()
        while i < ln:
            l, h = i+1, ln-1
            while l < h:
                val = nums[i] + nums[l] + nums[h]
                if target == val:
                    res = val
                    return res

                if abs(target - val) < min_diff:
                    min_diff = abs(target - val)        
                    res = val
                
                if val < target:
                    l += 1
                    continue
                if val > target:
                    h -= 1
                    continue
                while l < ln - 1 and nums[l] == nums[l+1]:
                    l += 1
                
                l += 1
                
            while i < ln-1 and nums[i] == nums[i+1]:
                i += 1
            i += 1
        return res
        
s = Solution()
print s.threeSumClosest([-1,2,1,-4], 1)        

File: go/src/leetcode/16. 3Sum Closest/16. 3Sum Closest.go
package main

import (
	"fmt"
	"math"
	"sort"
)

func threeSumClosest(nums []int, target int) int {
	minVal := math.MaxInt32
	ans := 0
	sort.Ints(nums)
	n := len(nums)
	for i := 0; i < n; i++ {
		l, r := i+1, n-1
		for l < r {
			val := nums[i] + nums[l] + nums[r]
			if val < target {
				l++
			} else if val > target {
				r--
			} else {
				return target
			}
			if math.Abs(float64(val - target)) < float64(minVal) {
				minVal = int(math.Abs(float64(val - target)))
				ans = val
			}
		}
	}

	return ans
}

func main() {
	fmt.Println(threeSumClosest([]int{-1,2,1,-4}, 1))
}